### PR TITLE
High-level-synthesis for permutations

### PR DESCRIPTION
**Pull request by alexanderivrii**
### Summary

This PR is based on top of #9082. First, it introduces`PermutationGate` for representing permutation logic, which is of type `Gate` rather than `QuantumCircuit`. The new way of storing permutation logic avoids synthesizing a permutation circuit when the ``Permutation`` gate is declared, delaying the actual synthesis to transpiler. It also allows to easily choose between several different algorithms for synthesizing permutations, which are available as high-level-synthesis permutation plugins.

Then it adds 3 different high-level-synthesis plugins for permutations. The ``BasicSynthesisPermutation`` plugin applies to fully-connected architectures and is based on sorting. This is the previously used algorithm for constructing quantum circuits for permutations. The ``ACGSynthesisPermutation`` plugin also applies to fully-connected architectures but is based on the Alon, Chung, Graham method. It synthesizes any permutation in depth 2 (measured in terms of SWAPs). The ``KMSSynthesisPermutation`` plugin applies to linear nearest-neighbor architectures and corresponds to the recently added Kutin, Moulton, Smithline method.

The ``PermutationGate`` has a new method ``__array__`` that returns a unitary matrix for a permutation, in particular allowing to create unitary matrices from circuits containing permutation gates. This might be immediately useful for #8597 (see https://github.com/Qiskit/qiskit-terra/pull/8597#discussion_r1023089761).

In the future, it would also make sense to add the permutation synthesis plugin based on ``AproximateTokenSwapper``.

### Example:

```
import numpy as np

from qiskit.circuit.library import PermutationGate
from qiskit.circuit import QuantumCircuit
from qiskit.transpiler.passes.synthesis.plugin import HighLevelSynthesisPluginManager

np.random.seed(1)

nq = 10
perm = PermutationGate(np.random.permutation(nq))
qc = QuantumCircuit(nq)
qc.append(perm, range(nq))

qct = PassManager(HighLevelSynthesis()).run(qc)
print(f"Default: {qct.size() = }, {qct.depth() = }")

qct = PassManager(HighLevelSynthesis(HLSConfig(permutation=[("kms", {})]))).run(qc)
print(f"kms: {qct.size() = }, {qct.depth() = }")

qct = PassManager(HighLevelSynthesis(HLSConfig(permutation=[("basic", {})]))).run(qc)
print(f"basic: {qct.size() = }, {qct.depth() = }")

qct = PassManager(HighLevelSynthesis(HLSConfig(permutation=[("acg", {})]))).run(qc)
print(f"acg: {qct.size() = }, {qct.depth() = }")
```

produces

```
Default: qct.size() = 7, qct.depth() = 5
kms: qct.size() = 21, qct.depth() = 8
basic: qct.size() = 7, qct.depth() = 5
acg: qct.size() = 7, qct.depth() = 2
```

Note that "kms" method adheres to LNN architecture. A quick experiment on more random permutations shows that the new "acq" method seems to produce the same number of gates as our previous method, but has better depth (which is always at most 2 compared to 5 in this example). However, I would like to add support for coupling map and run more experiments before changing the default.

### CODE REVIEW

**Comment by alexanderivrii** [comment]
I have added a few more tweaks required for fully turning a `Permutation` into a `Gate`. This PR should be ready for review (the docs seem to be failing for unrelated reasons).

Detailed comments:

We can add permutations to quantum circuits using the `permutation` method. As an example, `qc.permutation([1, 2, 0], [4, 0, 2])` means that the permutation with permutation pattern `[1, 2, 0]` is added onto qubits `[4, 0, 2]`. This is equivalent to `qc.append(Permutation(3, [1, 2, 0]), [4, 0, 2])`.

Comment related to the above. I have not changed the Permutation's `__init__` method, which accepts `num_qubits`, optional `pattern` and optional `seed`, and which creates a random permutation over `num_qubits` when the permutation pattern is `None`. Though I think it would be better to move the functionality of creating a random permutation to an external function and to simplify the `__init__` method to only accept the (non-trivial) permutation pattern. 

We had a long discussion on supporting `qasm()` for permutations, and as a short-term solution we have agreed on mimicking the solution for unitary gates (except that the name of a permutation gate can be defined uniquely based on the pattern). This is done in https://github.com/Qiskit/qiskit-terra/pull/9157/commits/ec71978943361703184bcf6dbf2c9a93ba80dbd9. Note that previously permutations were quantum circuits with names uniquely constructed from the permutation pattern, and now permutations are custom gates with name "Permutation" but `_qasm_name` uniquely constructed from the permutation pattern, yet the `qasms` should be exactly the same before and after the change.

Since permutations are gates, circuits with permutation gates naturally support methods like `c_if()` or `reverse_ops()`, and I had to add the method `inverse()` to Permutations to support taking inverses of circuits with permutation gates. There was also a minor tweak that needed to be done to support `qpy` for Permutations (the only reason is that `params` for Permutation only contain the pattern, but the `__init__` method requires the number of qubits and the pattern, see my comment before).

File: qiskit/synthesis/__init__.py [hunk 1]
```diff
@@ -77,8 +79,12 @@
     QDrift,
 )
 
+from .permutation import (
+    synth_permutation_depth_lnn_kms,
+    synth_permutation_basic,
+    synth_permutation_acg,
```
Comment (ShellyGarion) [inline qiskit/synthesis/__init__.py:85]: should we call it `synth_permutation_depth_acg` to emphasize that we optimize depth?

File: qiskit/synthesis/__init__.py [hunk 2]
```diff
@@ -77,8 +79,12 @@
     QDrift,
 )
 
+from .permutation import (
+    synth_permutation_depth_lnn_kms,
+    synth_permutation_basic,
```
Comment (ShellyGarion) [inline qiskit/synthesis/__init__.py:84]: Is `synth_permutation_basic` better than `synth_permutation_full`?
(in other methods we used "full")

File: test/python/synthesis/test_permutation_synthesis.py [hunk 3]
```diff
@@ -40,6 +42,42 @@ def test_get_ordered_swap(self, width):
...
+            for instruction in qc.data:
+                self.assertEqual(instruction.operation.name, "swap")
+
+            # Construct a linear function from the synthesized circuit, and
+            # check that its permutation pattern matches the original pattern.
+            synthesized_pattern = LinearFunction(qc).permutation_pattern()
+            self.assertTrue(np.array_equal(synthesized_pattern, pattern))
+
+    @data(4, 5, 10, 15, 20)
+    def test_synth_permutation_acg(self, width):
```
Comment (ShellyGarion) [inline test/python/synthesis/test_permutation_synthesis.py:64]: it's worth to add a test that the SWAP depth is 2, as well as to check that the number of SWAPs is the same as in the basic method

File: qiskit/circuit/library/generalized_gates/permutation.py [hunk 4]
```diff
@@ -77,17 +83,59 @@ def __init__(
...
-        name = "permutation_" + np.array_str(pattern).replace(" ", ",")
+        # This is needed to support qasm()
+        self._qasm_name = "permutation__" + "_".join([str(n) for n in pattern]) + "_"
+        self._qasm_definition = None
 
-        circuit = QuantumCircuit(num_qubits, name=name)
+        super().__init__(name="permutation", num_qubits=num_qubits, params=[pattern])
 
-        super().__init__(num_qubits, name=name)
+    def __array__(self, dtype=None):
```
Comment (ShellyGarion) [inline qiskit/circuit/library/generalized_gates/permutation.py:92]: Is there a test for this?

File: qiskit/circuit/library/generalized_gates/permutation.py [hunk 5]
```diff
@@ -77,17 +83,59 @@ def __init__(
...
+            # convert row to bitstring, reverse, apply permutation pattern, reverse again,
+            # and convert to row
+            bit = bin(r)[2:].zfill(nq)[::-1]
+            permuted_bit = "".join([bit[j] for j in self.pattern])
+            pr = int(permuted_bit[::-1], 2)
+            mat[pr, r] = 1
+
+        return mat
+
+    def validate_parameter(self, parameter):
```
Comment (ShellyGarion) [inline qiskit/circuit/library/generalized_gates/permutation.py:107]: why is this needed? what is the `parameter`?

File: qiskit/synthesis/__init__.py (hunk 1 above)
Comment (alexanderivrii) [reply qiskit/synthesis/__init__.py:87]: Thanks! I am afraid that we won't be able to find the perfect naming conventions. I am slightly in favor of keeping the current (shorter) variant of the name as there is only one "ACG" synthesis algorithm for permutations (the Alon, Chung, Graham method). In addition, while the ACG method does indeed guarantee the worst-case depth of 2 (measured in terms of SWAPs), it's also good for count, that's another reason why it might be better not to add "depth" to the name.

File: qiskit/synthesis/__init__.py (hunk 2 above)
Comment (alexanderivrii) [reply qiskit/synthesis/__init__.py:86]: I am happy to rename this, but I wanted to stress that this is the original algorithm we had in Qiskit rather than some yet another algorithm that works for all-to-all connectivity. Well, I am slightly in favor of keeping names shorter, so when a synthesis algorithms assumes special connectivity (such as LNN or grid), it may be useful to somehow represent this in the method's name, and if it does not - then just leave this out from the name. On the other hand, I am not sure how to name algorithms that work for any restricted connectivity. In any case, the name does capture the most important aspects of the routine: it's a synthesis algorithm for permutations. What do you think?

File: test/python/synthesis/test_permutation_synthesis.py (hunk 3 above)
Comment (alexanderivrii) [reply test/python/synthesis/test_permutation_synthesis.py:64]: Good point, adding this check.

File: qiskit/circuit/library/generalized_gates/permutation.py (hunk 4 above)
Comment (alexanderivrii) [reply qiskit/circuit/library/generalized_gates/permutation.py:92]: Good point. I have added the missing tests for this and other new functionality in [e826085](https://github.com/Qiskit/qiskit-terra/pull/9157/commits/e8260856117a83db1221eda62e3c1a654450ac89).

File: qiskit/circuit/library/generalized_gates/permutation.py (hunk 5 above)
Comment (alexanderivrii) [reply qiskit/circuit/library/generalized_gates/permutation.py:107]: This is unfortunately needed. `Permutation` inherit from `Gate` and `parameter` is whatever is in `params`, see this line in `Permutation.__init__`: `super().__init__(name="permutation", num_qubits=num_qubits, params=[pattern])`, i.e. for permutations the only parameter is the permutation pattern. The `Gate` class calls `validate_parameter` which does not like arrays, and the common strategy is to override this method in the child class (e.g., we do the same in `LinearFunction`).

**Review by ShellyGarion — DISMISSED** [review]

**Review by ShellyGarion — APPROVED** [review]
LGTM! thanks!

File: qiskit/circuit/library/generalized_gates/permutation.py [hunk 6]
```diff
@@ -16,20 +16,20 @@
...
 import numpy as np
 
-from qiskit.circuit.quantumcircuit import QuantumCircuit
+from qiskit.circuit.quantumcircuit import Gate
 from qiskit.circuit.exceptions import CircuitError
 
 
-class Permutation(QuantumCircuit):
-    """An n_qubit circuit that permutes qubits."""
+class Permutation(Gate):
```
Comment (mtreinish) [inline qiskit/circuit/library/generalized_gates/permutation.py:23]: I'm a bit concerned about making this change from a backwards compatibility PoV in a single release. While I agree with this being a `Gate` subclass instead of a `QuantumCircuit` is where we want to be I'm worried about any users of this class today that may be using it with a  circuit as is. For example the docs here show a good example of this with `Permutation(...).draw()`, but there are others I can think of the top of my head would be something like: `Permutation(...).compose(sub_circuit)` or `Permutation.measure_all()` all of which were previously completely valid which would be broken by this change.

 I'm thinking we need to figure out a better way to manage this migration which leaves the `QuantumCircuit` based `Permutation` class in place as is for now but establishes a new `Gate` based class as a standalone thing (like `PermutationGate` or something) in this PR and maybe update this `QuantumCircuit` based one to just wrap the new `Gate` class. Then in the next release we can deprecate the `QuantumCircuit` version and tell people to use the gate directly.

**Review by mtreinish — CHANGES_REQUESTED** [review]
I think the code itself LGTM, the `Gate` based class and synthesis routines look fine (although I scanned it kind of quickly). My bigger concern here though is around backwards compatibility. Making a breaking change to the `Permutation` class like this is a bit problematic and will likely break any existing users of the class. For this PR I feel we should introduce the `Gate` based permutation as a new class and update the circuit based class to internally just be a single `PermutationGate`.

Then to make this transition I think we'll have to follow this basic playbook:

1. This release add new `Gate` based permutation class (`PermutationGate` is the class name I can think of, but any other ideas are welcome)
2. Next release deprecate `Permutation` (`QuantumCircuit` based class)
3. 2 releases later remove `Permutation`
4. 3 releases later add alias of `PermutationGate` to old `Permutation` name
5. 4 releases later deprecate `PermutationGate`
6. 6 releases later remove `PermutationGate` in favor of `Gate` based `Permutation` class

4-6 are optional and only really needed if we strongly desire the end result of `Permutation` being a `Gate` (and not having a new name)

File: qiskit/circuit/library/generalized_gates/permutation.py (hunk 6 above)
Comment (alexanderivrii) [reply qiskit/circuit/library/generalized_gates/permutation.py]: Thanks, this is a very good point. I keep relying on qiskit-terra tests to make sure that nothing gets broken, yet I keep forgetting that there is also a lot of other code outside of qiskit-terra. The example with `Permutation.measure_all()` clearly shows why we need to keep the old functionality. It should be quite straightforward for me to keep the old implementation and to rename the old implementation, in this way this development no longer changes any existing behavior, it only adds new things.

I have already started doing changes, but here are a couple of preliminary thoughts. The name `PermutationGate` for the new class is good. For objects of type `PermutationGate` we can still have `op.name = "permutation"`, since there is no danger to confuse permutation circuits and permutation gates (and in any case the name assigned to a permutation circuit is of the form `permutation_[2,4,3,0,1]`). In particular, we can keep using names like `permutation.kms` in `entry_points` to specify permutation synthesis algorithms for `PermutationGate`.

The suggested deprecation flow also allows to avoid passing `num_qubits` and `seed` to the initializer of the `PermutationGate`, making the code a bit cleaner. There is a function `np.random.permutation` that returns a random permutation pattern, from which a `PermutationGate` can be constructed.

File: qiskit/circuit/quantumcircuit.py [hunk 7]
```diff
@@ -4077,6 +4087,21 @@ def pauli(
 
         return self.append(PauliGate(pauli_string), qubits, [])
 
+    def permutation(self, pattern: List[int], qubits: Sequence[QubitSpecifier]) -> InstructionSet:
```
Comment (mtreinish) [inline qiskit/circuit/quantumcircuit.py:4090]: Do we really need a circuit method for the permutation like this? We typically only do this for standard gates, is there a pressing need to do this? If not I'd rather defer this for now, if the need arises later it's easy to add. But, it's hard to roll this back if we decided it's not really useful in the future.

File: qiskit/circuit/library/generalized_gates/permutation.py [hunk 8]
```diff
@@ -88,3 +92,112 @@ def __init__(
 
         all_qubits = self.qubits
         self.append(circuit.to_gate(), all_qubits)
+
+
+class PermutationGate(Gate):
```
Comment (mtreinish) [inline qiskit/circuit/library/generalized_gates/permutation.py:97]: Do you want to use this inside the `Permutation` class's definition. For example:

```python3
class Permutation(QuantumCircuit):

    def __init__(
        self,
        num_qubits: int,
        pattern: Optional[List[int]] = None,
        seed: Optional[int] = None,
    ) -> None:
        if pattern is not None:
            if sorted(pattern) != list(range(num_qubits)):
                raise CircuitError(
                    "Permutation pattern must be some ordering of 0..num_qubits-1 in a list."
                )
            pattern = np.array(pattern)
        else:
            rng = np.random.default_rng(seed)
            pattern = np.arange(num_qubits)
            rng.shuffle(pattern)
        circuit = QuantumCircuit(num_qubits, name=name)
        super().__init__(num_qubits, name=name)
        circuit.append(PermutationGate(pattern), self.qubits)
```

Or do you think the difference in behavior is too much?

File: qiskit/circuit/quantumcircuit.py [hunk 9]
```diff
@@ -1698,20 +1698,30 @@ def qasm(
...
-
-                    existing_composite_circuits.append(operation)
-                    _add_sub_instruction_to_existing_composite_circuits(
-                        operation, existing_gate_names, existing_composite_circuits
-                    )
+                if operation.name not in existing_gate_names:
+                    # If the operation has a unique qasm name, we set the name of the operation
+                    # to this qasm name. This is now the case for the permutation gate, for which
+                    # the unique qasm name is based on the permutation pattern.
+                    op_qasm_name = getattr(operation, "_qasm_name", None)
```
Comment (mtreinish) [inline qiskit/circuit/quantumcircuit.py:1705]: Do we document this anywhere, like in the operation class or something? I think this might come as a surprise if I were to implement a new gate and wanted to serialize the qasm in a similar manner to permutation.

File: test/python/circuit/library/test_permutation.py [hunk 10]
```diff
@@ -39,5 +42,125 @@ def test_permutation_bad(self):
...
+        self.assertTrue(np.array_equal(qc.data[0].operation.pattern, qcr.data[0].operation.pattern))
+        self.assertTrue(np.array_equal(qc.data[0].qubits, qcr.data[0].qubits))
+
+    def test_conditional(self):
+        """Test adding conditional permutations."""
+        qc = QuantumCircuit(5, 1)
+        qc.permutation([1, 2, 0], [2, 3, 4]).c_if(0, 1)
+        self.assertIsNotNone(qc.data[0].operation.condition)
+
+    def test_qasm(self):
```
Comment (mtreinish) [inline test/python/circuit/library/test_permutation.py:145]: Can you add a similar test for qpy serialization. Looking at the gate definition I don't expect any issues since the permutation pattern is defined in the gate parameters so qpy should just pass the list of ints to the constructor and everything should work fine, but it'd be good to validate it is recreating it exactly. (it'd  be fine to add here or in `test/python/qpy`)

File: qiskit/circuit/quantumcircuit.py (hunk 7 above)
Comment (alexanderivrii) [reply qiskit/circuit/quantumcircuit.py:4090]: There is definitely no pressing need for this, I thought that the convention was to add a method to QuantumCircuit for every object that inherits from Gate; sure, I can remove it.

File: qiskit/circuit/library/generalized_gates/permutation.py (hunk 8 above)
Comment (alexanderivrii) [reply qiskit/circuit/library/generalized_gates/permutation.py:97]: This is an interesting suggestion. It would probably work, but (as the PermutationGate has no definition) it could also in principle break something (like it happened for Cliffords). I did check and added tests for methods like `inverse`, `reversed_ops`, etc. there might be other circuit's methods that I have not fully considered. So maybe for now let's err on the safe side and keep the older implementation as is?

File: qiskit/circuit/quantumcircuit.py (hunk 9 above)
Comment (alexanderivrii) [reply qiskit/circuit/quantumcircuit.py:1705]: Handling qasm for permutation gates is very heavily based on the code for unitary gates, except for the idea of assigning a name based on the pattern (which is already done for permutation circuits) without using an auxiliary id. What is your recommendation I should do here?

File: test/python/circuit/library/test_permutation.py (hunk 10 above)
Comment (alexanderivrii) [reply test/python/circuit/library/test_permutation.py:148]: Done in [e577372](https://github.com/Qiskit/qiskit-terra/pull/9157/commits/e57737220bb6adeebc7c8fcd111943b88f3920ce)

File: qiskit/circuit/quantumcircuit.py (hunk 7 above)
Comment (alexanderivrii) [reply qiskit/circuit/quantumcircuit.py]: Done in [e1e9f05](https://github.com/Qiskit/qiskit-terra/pull/9157/commits/e1e9f05de3615f2fc9b2ec55696061304b9d4ef8)

File: qiskit/circuit/quantumcircuit.py (hunk 9 above)
Comment (mtreinish) [reply qiskit/circuit/quantumcircuit.py:1705]: Well it's more that this is an implicit API we're adding that if an `Operation` instance has a `_qasm_name` attribute that attribute is now used to specify the name of the gate when generating openqsm2. This is a new interface and we're not documenting it anywhere. I'm also a bit concerned about being tied to this moving forward as it kind of couples the operation object to its qasm representation which has been a source of problems with the `Instruction.qasm()` method (although this is much looser than the problems around `Instruction.qasm()`).

That being said I think maybe if we just leave this as a hack in place for `PermutationGate` for now that's enough so maybe something like:

```suggestion
                    if operation.name == "permutation":
                        op_qasm_name = getattr(operation, "_qasm_name")
```

(or maybe an `isinstance` check)

Comment (alexanderivrii) [reply qiskit/circuit/quantumcircuit.py]: Thanks for the explanation, and I agree, this is a cleaner "hack" that what I had. I also agree that we want a clean way (without hidden implicit assumptions) to specify qasm2 names for high-level-objects, which is somehow also in the spirit of other things we discussed: inverses, reversed_ops and so on.

**Review by mtreinish — APPROVED** [review]
This LGTM now, thanks for doing this and making all the updates from review.
